sudoku_solver/draft_one/lib/sudoku/puzzle.rb

```diff
@@ -0,0 +1,269 @@
+require 'matrix'
+
+require_relative '../n_tree'
+require_relative '../doubly_linked_list'
+
+module Sudoku
+
+  class Puzzle
+    include NTree
+    include DoublyLinkedList
+
+    @@ROWS = [(0..8),
+              (9..17),
+              (18..26),
+              (27..35),
+              (36..44),
+              (45..53),
+              (54..62),
+              (63..71),
+              (72..80)]
+
+    attr_accessor :parent, :children, :next, :prev
+    attr_reader :head, :tail, :last_sibling
+    # @param [Array<Integer>] blocks
+    # @param [Puzzle] parent
+    # @param [Puzzle] children
+    # @param [Puzzle] nxt
+    # @param [Puzzle] prev
+    # @param [Puzzle] head
+    # @param [Puzzle] tail
+    def initialize(blocks, parent = nil, children = nil, nxt = nil, prev = nil, head = nil, tail = nil)
```

Comment:
We've now translated the problem of solving Sudoku into the problem of implementing a depth-first search of this tree.  The latter problem doesn't require a doubly linked list, which is why I'm not clear on its role in your code.